### PR TITLE
Add greenPowerKey for ZGP devices

### PR DESCRIPTION
**Pull request by fredericseiler**
Issue Koenkk/zigbee-herdsman#398

This PR adds a `greenPowerKey` fingerprint filter to better identify ZGP devices.

See main PR (Koenkk/zigbee-herdsman#403) for more details.

Also adds support for:
- EnOcean PTM 215Z
- EnOcean PTM 215ZE
- Legrand ZLGP15
- Legrand ZLGP17 / ZLGP18

### CODE REVIEW

**Comment by fredericseiler** [comment]
Note : the PTM215Z and PTM215ZE converters are not complete: there are still some actions missing like energy bar press/release only (without pressing a button) or combinations that can only be triggered when you disassemble the switch (not common use cases). For the sake of completeness, I'll add those later when I'll have more bandwidth.

File: devices/enocean.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            {modelID: 'GreenPower_2', greenPowerKey: '0.46.197.242'}],
+        model: 'PTM 215Z',
+        vendor: 'EnOcean',
+        description: 'Pushbutton transmitter module',
+        fromZigbee: [fz.enocean_ptm215z],
+        toZigbee: [],
+        exposes: [e.action(['press_1', 'release_1', 'press_2', 'release_2', 'press_3', 'release_3', 'press_4', 'release_4',
+            'press_1_and_3', 'release_1_and_3', 'press_2_and_4', 'release_2_and_4'])],
+        whiteLabel: [
+            {vendor: 'Niko', description: 'Dimmer switch for Hue system', model: '91004'},
```
Comment (Koenkk) [inline devices/enocean.js:17]: @fredericseiler made some updates, added the `PTM 216Z`. Suprisingly my Niko 91004 had a different `greenPowerKey` (`0.46.197.242`). Did you test this with different devices with the same model?

**Comment by danieledwardgeorgehitchcock** [comment]
Hi all, would I need to re-test my Vimar converter (used to be bundled in the greenpower_2 device once this has been merged?

**Comment by Koenkk** [comment]
@danieledwardgeorgehitchcock it would be great if you could test this is indeed, how are you running z2m? (bare metal/docker?)

**Comment by danieledwardgeorgehitchcock** [comment]
I'm running in Docker

**Comment by Koenkk** [comment]
Can you try:

```sh
cd /opt
git clone https://github.com/fredericseiler/zigbee-herdsman.git -b zgp
cd zigbee-herdsman
npm ci
npm run build
```

Now map the volume via e.g. docker run command: `-v /opt/zigbee-herdsman:/app/node_modules/zigbee-herdsman`, force remove your green power device, pair it again and share the resulting data/database.db entry here.

**Comment by danieledwardgeorgehitchcock** [comment]
Hi @Koenkk - I have followed your guide above and built Zigbee Herdsman and have mapped the built directory into my Docker container however, I am not seeing any of the GreenPower attributes that are seen above so, I am not sure if the process has worked. Regardless here is the database.db entry for the Vimar:

```json
{"id":2,"type":"GreenPower","ieeeAddr":"0x000000000170ccdd","nwkAddr":52445,"manufId":null,"modelId":"GreenPower_2","epList":[242],"endpoints":{"242":{"epId":242,"inClusterList":[],"outClusterList":[],"clusters":{},"binds":[],"configuredReportings":[],"meta":{}}},"interviewCompleted":true,"meta":{},"lastSeen":1629117995835}
```

Here is the one for the Hue Tap:
```json
{"id":3,"type":"GreenPower","ieeeAddr":"0x0000000000432331","nwkAddr":9009,"manufId":null,"modelId":"GreenPower_2","epList":[242],"endpoints":{"242":{"epId":242,"inClusterList":[],"outClusterList":[],"clusters":{},"binds":[],"configuredReportings":[],"meta":{}}},"interviewCompleted":true,"meta":{},"lastSeen":1629118050602}
```

**Comment by fredericseiler** [comment]
@danieledwardgeorgehitchcock what's the model name of the EnOcean module behind your Vimar switch?

`PTM 215Z` or `PTM 216Z`?

**Comment by danieledwardgeorgehitchcock** [comment]
@fredericseiler this looks like a whitelabel EnOcean product so, there is no distinguishable identifiers on the device other than the ones that Vimar have added

**Comment by fredericseiler** [comment]
2021-08-16-16-49-31 Preview 2021-08-16-16-48-29 CleanShot X CleanShot png

**Comment by danieledwardgeorgehitchcock** [comment]
D'oh! yes confirmed! Sorry!

**Comment by fredericseiler** [comment]
Well, it's reassuring in one way because a PTM 216Z should be identified as `GreenPower_7` (Device Type 0x07).

Now for the bad news: `dataPayload.frame.Payload.options` apparently can't be trusted.

I can't find any PTM 215Z recent user manual (> 2014) with comprehensive details about its ZGP commissioning telegram (unlike the PTM 216Z), and it's really annoying.

**Comment by danieledwardgeorgehitchcock** [comment]
> Well, it's reassuring in one way because a PTM 216Z should be identified as `GreenPower_7` (Device Type 0x07).
> 
> Now for the bad news: `dataPayload.frame.Payload.options` apparently can't be trusted.
> 
> I can't find any PTM 215Z recent user manual (> 2014) with comprehensive details about its ZGP commissioning telegram (unlike the PTM 216Z), and it's really annoying.

This any use?:

https://www.enocean.com/en/products/enocean_modules_24ghz/ptm-215ze/user-manual-pdf/

**Comment by fredericseiler** [comment]
Nop, it's the PTM215Z**E**.

**Comment by danieledwardgeorgehitchcock** [comment]
Deffo the PTM 215Z:

IMG_4746

**Comment by fredericseiler** [comment]
Thanks for the pic.

@Koenkk So... We can throw away all this greenPowerKey stuff, right? 😢

**Comment by fredericseiler** [comment]
@Koenkk in the ZCL payload, which of the two "options" is the one from the ZGP specs? (sorry, having a hard time trying to understand `zcl/zclFrame.ts` and `deconz/driver/frameParser.ts`)

```json5
{
    "frame": {
        // ...
        "Payload": {
            "options": 62149,
            // ...
            "payloadSize": 37,
            "commandFrame": {
                "deviceID": 254,
                "options": 197,
                "extendedOptions": 242,
                // ...
            }
        },
        // ...
}
```

Specs:
2021-08-17-13-01-31 Google Chrome ZigBee Green Power Specification

**Comment by Koenkk** [comment]
ZGP spec = `zcl/zclFrame.ts` (so the `commandFrame` is figure 71)

**Comment by fredericseiler** [comment]
Where does the first "options" (62149 in my example) come from?

**Comment by fredericseiler** [comment]
I just bought another PTM 215Z (NodOn CWS-4-1-01_HUE):

```json5
{
    "frame": {
        // ...
        "Payload": {
            "options": 62149,
            // ...
            "payloadSize": 46,
            "commandFrame": {
                "deviceID": 2,
                "options": 197,
                "extendedOptions": 242,
                // ...
            }
        },
        // ...
}
```

I'll exclude the first `options` for now (has it something to do with the user setup/coordinator?). Also, `extendedOptions` has always the same value, so I'll exclude it too.

Now, the green power key is basically: `payloadSize` `.` `commandFrame.options`.

- `27.129`: EnOcean PTM 215ZE (Trio2sys T2S20040004)
- `31.197`: Legrand ZLGP17 / ZLGP18
- `46.197`: EnOcean PTM 215Z (Niko 91004 / Vimar 03906 / NodOn CWS-4-1-01_HUE)

Unfortunately, I have no idea why @danieledwardgeorgehitchcock's Hue Tap has the same key as the PTM 215ZE (`27.129`). An older version? Or a specific one for Philips/Signify?

@Koenkk is it worth continuing in this direction?

**Comment by Koenkk** [comment]
@fredericseiler the options come from https://github.com/Koenkk/zigbee-herdsman/blob/145abce96a7d8e8814a2668b55ddb066439eec43/src/zcl/definition/cluster.ts#L1199 , wireshark shows the following for this (Hue tap in this case)

image

**Comment by danieledwardgeorgehitchcock** [comment]
> I just bought another PTM 215Z (NodOn CWS-4-1-01_HUE):
> 
> ```js
> {
>     "frame": {
>         // ...
>         "Payload": {
>             "options": 62149,
>             // ...
>             "payloadSize": 46,
>             "commandFrame": {
>                 "deviceID": 2,
>                 "options": 197,
>                 "extendedOptions": 242,
>                 // ...
>             }
>         },
>         // ...
> }
> ```
> 
> I'll exclude the first `options` for now (has it something to do with the user setup/coordinator?). Also, `extendedOptions` has always the same value, so I'll exclude it too.
> 
> Now, the green power key is basically: `payloadSize` `.` `commandFrame.options`.
> 
> * `27.129`: EnOcean PTM 215ZE (Trio2sys T2S20040004)
> * `31.197`: Legrand ZLGP17 / ZLGP18
> * `46.197`: EnOcean PTM 215Z (Niko 91004 / Vimar 03906 / NodOn CWS-4-1-01_HUE)
> 
> Unfortunately, I have no idea why @danieledwardgeorgehitchcock's Hue Tap has the same key as the PTM 215ZE (`27.129`). An older version? Or a specific one for Philips/Signify?
> 
> @Koenkk, is it worth continuing in this direction?

It will probably be an older one.. My Hue Tap is ancient!

**Comment by fredericseiler** [comment]
@Koenkk Damn, I have a different structure in Wireshark, without `options`!

2021-08-18-18-52-04 Wireshark Wi-Fi en0

I guess that's because I use a Conbee 2 (with the [Zshark](https://phoscon.de/en/conbee/software#zshark) app and firmware), so I'll have to figure out another way to see what's inside `"options": 62149`.

**Comment by fredericseiler** [comment]
@Koenkk, in my case, `62149` corresponds to `commandFrame.options` and `commandFrame.extendedOptions`:

`62149` = `0xF2C5` = `1111 0010 1100 0101`

2021-08-18-21-35-51 Wireshark Wi-Fi en0

Am I missing something?

**Comment by Koenkk** [comment]
@fredericseiler this should split them: https://github.com/Koenkk/zigbee-herdsman/pull/403/commits/840dd766b71ad94b522a1ac386a1d73855363998

**Comment by fredericseiler** [comment]
I was wondering why we previously got a different value for `Payload.options`: `62149` for me, `0` for you and `2048` for @danieledwardgeorgehitchcock, because if `Payload.options` is just a merge of `Payload.commandFrame.options` and `Payload.commandFrame.extendedOptions` (which are the same for each of us, respectively `197` and `242`), then `Payload.options` won't be different, right ?

Anyway, here's the new ZCL frame for a PTM 215Z:

```json5
{
    "frame": {
        "Payload": {
            "options": 197,
            "extendedOptions": 242,
            // ..
            "payloadSize": 46,
            "commandFrame": {
                "deviceID": 2,
                "options": 197,
                "extendedOptions": 242,
                // ...
            }
        }
    }
}
```

@Koenkk, can you confirm it with your own devices?

And now that we can only use `Payload.options` and `Payload.payloadSize` to identify a ZGP device, isn't it easier to just add them both as fingerprint attributes, instead of the whole greenPowerKey stuff?

```js
module.exports = [
    {
        fingerprint: [{modelID: 'GreenPower_2', payloadOptions: '197', payloadSize: '46'}],
    }
];
```

**Comment by Koenkk** [comment]
@fredericseiler isn't enough to just drop the `dataPayload.frame.Payload.options` from the `greenPowerKey`? With this my `PTM 215Z` has the same `greenPowerKey` as yours. My Hue tap, PTM 216Z and PTM 215Z get recognised now with the current changes.

@danieledwardgeorgehitchcock could you also test with the latest changes in this pr and https://github.com/Koenkk/zigbee-herdsman/pull/403 ?

**Comment by fredericseiler** [comment]
> isn't enough to just drop the dataPayload.frame.Payload.options from the greenPowerKey?

Ok for me.

> interestingly, the PTM 215ZE and hue tap have the same greenPowerKey now. UPDATE: Hue tap actually contains a PTM 215Z, should we merge the PTM 215ZE and PTM 215Z and add the Hue tap as a whitelabel?

Beware, PTM 215ZE (`27.129.242`) != PTM 215Z (`46.197.242`) (also the commandIDs are [not the same](https://github.com/Koenkk/zigbee-herdsman-converters/blob/47569ac78976bb96171efcb79a22ab78d1ea431a/converters/fromZigbee.js#L3712-L3762))

As for the Hue Tap, on every pic I found about the inner model, it's always a PTM 215Z, so I guess we can whitelabel it.

What I don't understand is why the module tagged "PTM 215Z" in @danieledwardgeorgehitchcock's Hue Tap has the same fingerprint as a PTM 215ZE. Perhaps, with luck, it also shares the same commandIDs as the PTM 215ZE (and then we can confirm that it's an older version, as already discussed here).

@Koenkk Any thoughts on this? I'm still perplexed about it:
> I was wondering why we previously got a different value for Payload.options: 62149 for me, 0 for you and 2048 for @danieledwardgeorgehitchcock, because if Payload.options is just a merge of Payload.commandFrame.options and Payload.commandFrame.extendedOptions (which are the same for each of us, respectively 197 and 242), then Payload.options won't be different, right ?

**Comment by Koenkk** [comment]
> What I don't understand is why the module tagged "PTM 215Z" in @danieledwardgeorgehitchcock's Hue Tap has the same fingerprint as a PTM 215ZE. Perhaps, with luck, it also shares the same commandIDs as the PTM 215ZE (and then we can confirm that it's an older version, as already discussed here).

My Hue tap also has the same id as the PTM 215ZE 

> Any thoughts on this? I'm still perplexed about it:

`Payload.options` != `Payload.commandFrame.options`, I don't know why it is actually different (couldn't find good docs for it, maybe it depends on the  proxy it paired via (router)), but with Wireshark I confirmed that `Payload.options` `0` was sent.

**Comment by fredericseiler** [comment]
> My Hue tap also has the same id as the PTM 215ZE
@Koenkk By "id", you mean Device Type (2) or greenPowerKey (27.129.242 for the PTM 215ZE) ?

**Comment by Koenkk** [comment]
Sorry, I mean the fingerprint:

- Hue tap: https://github.com/Koenkk/zigbee-herdsman-converters/pull/2885/files#diff-d364da3112ff528cb40942e5c0b837ea0db3c5f46b0f5f99b29221fa6539d4e9R33
- PTM 215ZE: https://github.com/Koenkk/zigbee-herdsman-converters/pull/2885/files#diff-4c4b285c3ce53d300b8d2b1f5165d1e8c37746fedf7faa9c9497e43bffdd8df8R22

**Comment by fredericseiler** [comment]
@Koenkk What are the values when you press and release the button 1 (A0) on the Hue Tap?

1. 0x10 (16) when pressed, 0x14 (20) when released
2. 0x22 (34) when pressed, 0x23 (35) when released

**Comment by fredericseiler** [comment]
@Koenkk Nevermind, I've just seen the lookup table you built for the Hue Tap. To summarize:

PTM 215Z in FoH switches (46.197.242)
- Press 1: 0x10 (16)
- Release 1: 0x14 (20)
- Press 2: 0x11 (17)
- Release 2: 0x15 (21)
- Press 3: 0x13 (19)
- Release 3: 0x17 (23)
- Press 4: 0x12 (18)
- Release 4: 0x16 (22)

PTM 215ZE (27.129.242)
- Press 1: 0x22 (34)
- Release 1: 0x23 (35)
- Press 2: 0x18 (24)
- Release 2: 0x19 (25)
- Press 3: 0x14 (20)
- Release 3: 0x15 (21)
- Press 4,: 0x12 (18)
- Release 4: 0x13 (19)

PTM 215Z in Hue Tap (27.129.242 - same as 215ZE)
- Press 1: 0x22 (34) - same as "Press 1" from 215ZE
- Release 1: ?
- Press 2: 0x10 (16) - same as "Press 1" from 215Z
- Release 2: ?
- Press 3: 0x11 (17) - same as "Press 2" from 215Z
- Release 3: ?
- Press 4: 0x12 (18) - same as "Press 4" from 215ZE
- Release 4: ?

Maybe the Hue Tap shell buttons (1 dot, 2 dots, 3 dots, 4 dots) does not correspond to the module buttons (A0, A1, B0, B1). Let's refer to the PTM 215ZE user manual:

- Press 1 dot: 0x22 (34) => "Press energy bar + A0"
- Press 2 dots: 0x10 (16) => "Press energy bar alone"
- Press 3 dots: 0x11 (17) => "**Release** energy bar alone"
- Press 4 dots: 0x12 (18) => "Press energy bar + B1"

Well, a press action can't be a release event. Furthermore, the commissioning procedure for the Hue Tap is the same as the PTM 215Z (long press), not the PTM 215ZE (short press, long press, short press).

If we refer to the ZGP specs:

- Press 1 dot: 0x22 (34) => "Toggle"
- Press 2 dots: 0x10 (16) => "Scene 0"
- Press 3 dots: 0x11 (17) => "Scene 1"
- Press 4 dots: 0x12 (18) => "Scene 2"

So, my best guess is that the PTM 215Z found in the Hue Tap is another beast with its own commandIDs, which also unfortunately shares the same ZCL frame as the PTM 215ZE.

If you read the full model ID printed on the label of each device (where **X** is the revision number):

- PTM 215Z in FoH switches: S3071-A215 FOH-**X** DA
- PTM 215Z in Hue Tap: S3071-A215-**X** DA
- PTM 215ZE: S3271-A215-**X** DA

If there's no other way to differentiate them, I'm afraid that the greenPowerKey won't be of much help in that case.

Do you know someone working at EnOcean? 😅

**Comment by danieledwardgeorgehitchcock** [comment]
> @Koenkk Nevermind, I've just seen the lookup table you built for the Hue Tap. To summarize:
> 
> PTM 215Z in FoH switches (46.197.242)
> 
> * Press 1: 0x10 (16)
> * Release 1: 0x14 (20)
> * Press 2: 0x11 (17)
> * Release 2: 0x15 (21)
> * Press 3: 0x13 (19)
> * Release 3: 0x17 (23)
> * Press 4: 0x12 (18)
> * Release 4: 0x16 (22)
> 
> PTM 215ZE (27.129.242)
> 
> * Press 1: 0x22 (34)
> * Release 1: 0x23 (35)
> * Press 2: 0x18 (24)
> * Release 2: 0x19 (25)
> * Press 3: 0x14 (20)
> * Release 3: 0x15 (21)
> * Press 4,: 0x12 (18)
> * Release 4: 0x13 (19)
> 
> PTM 215Z in Hue Tap (27.129.242 - same as 215ZE)
> 
> * Press 1: 0x22 (34) - same as "Press 1" from 215ZE
> * Release 1: ?
> * Press 2: 0x10 (16) - same as "Press 1" from 215Z
> * Release 2: ?
> * Press 3: 0x11 (17) - same as "Press 2" from 215Z
> * Release 3: ?
> * Press 4: 0x12 (18) - same as "Press 4" from 215ZE
> * Release 4: ?
> 
> Maybe the Hue Tap shell buttons (1 dot, 2 dots, 3 dots, 4 dots) does not correspond to the module buttons (A0, A1, B0, B1). Let's refer to the PTM 215ZE user manual:
> 
> * Press 1 dot: 0x22 (34) => "Press energy bar + A0"
> * Press 2 dots: 0x10 (16) => "Press energy bar alone"
> * Press 3 dots: 0x11 (17) => "**Release** energy bar alone"
> * Press 4 dots: 0x12 (18) => "Press energy bar + B1"
> 
> Well, a press action can't be a release event. Furthermore, the commissioning procedure for the Hue Tap is the same as the PTM 215Z (long press), not the PTM 215ZE (short press, long press, short press).
> 
> If we refer to the ZGP specs:
> 
> * Press 1 dot: 0x22 (34) => "Toggle"
> * Press 2 dots: 0x10 (16) => "Scene 0"
> * Press 3 dots: 0x11 (17) => "Scene 1"
> * Press 4 dots: 0x12 (18) => "Scene 2"
> 
> So, my best guess is that the PTM 215Z found in the Hue Tap is another beast with its own commandIDs, which also unfortunately shares the same ZCL frame as the PTM 215ZE.
> 
> If you read the full model ID printed on the label of each device (where **X** is the revision number):
> 
> * PTM 215Z in FoH switches: S3071-A215 FOH-**X** DA
> * PTM 215Z in Hue Tap: S3071-A215-**X** DA
> * PTM 215ZE: S3271-A215-**X** DA
> 
> If there's no other way to differentiate them, I'm afraid that the greenPowerKey won't be of much help in that case.
> 
> Do you know someone working at EnOcean? 😅

Have you emailed someone at EnOcean?

It may be worth while reaching out to their support team to see if they can help out as a last ditch attempt.

I did the same with Innr and they were more than happy to help out as the Home Assistant scene (and therefore Zigbee2MQTT) is one of their biggest markets and wider support = more sales = good business sense!

**Comment by Koenkk** [comment]
I've marked the Hue tap as a whitelabel of the PTM 215ZE, combined with the added actions: https://github.com/Koenkk/zigbee-herdsman-converters/pull/2885/commits/0e9a56ded77efc38d088783eb2c9e9efd0dbbf7b#diff-b676ab0df4d49b8572967cb4c92a2dc6264091131f6b1828ca13d0f9656e367aR3756 it works perfectly now producing `press_1`, `press_2`, `press_3` and `press_4` actions which match the visual markings

**Comment by fredericseiler** [comment]
@Koenkk what about the other events from the Hue Tap (releases, and pressing/releasing two buttons at the same time)?

**Comment by Koenkk** [comment]
> @Koenkk what about the other events from the Hue Tap (releases, and pressing/releasing two buttons at the same time)?

I couldn't manage to trigger those

**Comment by fredericseiler** [comment]
@Koenkk One last thing, do you want a documentation & images PR for those devices (EnOcean PTM-* and Legrand ZLGP*)?

**Comment by Koenkk** [comment]
@fredericseiler would be great if you could do this, you can take pairing instructions from https://www.zigbee2mqtt.io/devices/GreenPower_On_Off_Switch.html . On the legacy green power devices you can add a note that this is deprecated and that you need to force remove + repair your device.

**Pull request merged**